packages/api-fetch/src/index.js

```diff
@@ -0,0 +1,69 @@
+/**
+ * Internal dependencies
+ */
+import createNonceMiddleware from './middlewares/nonce';
+import createRootURLMiddleware from './middlewares/root-url';
+import createPreloadingMiddleware from './middlewares/preloading';
+import namespaceEndpointMiddleware from './middlewares/namespace-endpoint';
+import httpV1Middleware from './middlewares/http-v1';
+
+const middlewares = [];
+
+function registerMiddleware( middleware ) {
+	middlewares.push( middleware );
+}
+
+function apiFetch( options ) {
+	const raw = ( nextOptions ) => {
+		const { url, path, body, data, parse = true, ...remainingOptions } = nextOptions;
+		const headers = remainingOptions.headers || {};
+		if ( ! headers[ 'Content-Type' ] && data ) {
+			headers[ 'Content-Type' ] = 'application/json';
+		}
+
+		const responsePromise = window.fetch(
+			url || path,
+			{
+				...remainingOptions,
+				credentials: 'include',
+				body: body || JSON.stringify( data ),
+				headers,
+			}
+		);
+		const checkStatus = ( response ) => {
+			if ( response.status >= 200 && response.status < 300 ) {
+				return response;
+			}
+
+			throw response;
+		};
+
+		const parseResponse = ( response ) => {
+			return parse ? response.json().catch( () => response ) : response;
+		};
+
+		return responsePromise
+			.then( checkStatus )
+			.then( parseResponse );
+	};
+	const steps = [
+		...middlewares,
+		namespaceEndpointMiddleware,
+		httpV1Middleware,
+		raw,
+	].reverse();
```

Comment:
I guess this is inherited, but why do we go through the effort to `Array#reverse` if we could already know what the result will be in assigning this array? i.e.

```js
const steps = [
	raw,
	httpV1Middleware,
	namespaceEndpointMiddleware,
	...middlewares,
];
```